pallets/score/src/lib.rs

```diff
@@ -0,0 +1,268 @@
+		/// A new journal entry has been added.
+		/// \[entry identifier, entity, author\]
+		JournalEntry {
+			identifier: JournalIdentifierOf,
+			entity: EntityIdentifierOf<T>,
+			author: RequestorIdentifierOf<T>,
+		},
+		/// Aggregate scores has been updated.
+		/// \[entity identifier\]
+		AggregateUpdated { entity: EntityIdentifierOf<T> },
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		/// Invalid Identifer Length
+		InvalidIdentifierLength,
+		// Invalid digest
+		InvalidDigest,
+		// Invalid creator signature
+		InvalidSignature,
+		// Invalid Rating Identifier
+		InvalidRatingIdentifier,
+		// Transaction already rated
+		TransactionAlreadyRated,
+		// Invalid rating value - should be between 1 and 50
+		InvalidRatingValue,
+		// Exceeds the maximum allowed entries in a single transaction
+		TooManyJournalEntries,
+		// Invalid entity signature
+		InvalidEntitySignature,
+		//Stream digest is not unique
+		DigestAlreadyAnchored,
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		///
+		/// Create a new rating identifier and associates it with its
+		/// controller. The controller (issuer) is the owner of the identifier.
+		///
+		/// * origin: the identity of the Transaction Author. Transaction author
+		///   pays the transaction fees
+		/// * tx_journal: the incoming rating entry.
+		#[pallet::weight(<T as pallet::Config>::WeightInfo::entries())]
+		pub fn entries(origin: OriginFor<T>, journal: JournalInputOf<T>) -> DispatchResult {
+			let author = <T as Config>::EnsureOrigin::ensure_origin(origin)?;
+
+			ensure!(
+				(journal.entry.score >= T::MinScoreValue::get()
+					&& journal.entry.score <= T::MaxScoreValue::get()),
+				Error::<T>::InvalidRatingValue
+			);
+			ensure!(
+				!<JournalHashes<T>>::contains_key(&journal.digest),
+				Error::<T>::DigestAlreadyAnchored
+			);
+
+			ensure!(
+				journal.signature.verify(&(&journal.digest).encode()[..], &journal.entry.entity),
+				Error::<T>::InvalidEntitySignature
+			);
+
+			let identifier = JournalIdentifierOf::try_from(
+				ss58identifier::generate(&(&journal.digest).encode()[..], SCORE_PREFIX)
+					.into_bytes(),
+			)
+			.map_err(|()| Error::<T>::InvalidIdentifierLength)?;
+
+			ensure!(
+				!<TidEntries<T>>::contains_key(&journal.entry.tid, &journal.entry.score_type),
+				Error::<T>::TransactionAlreadyRated
+			);
+
+			let block_number = frame_system::Pallet::<T>::block_number();
+
+			<Journal<T>>::insert(
+				&identifier,
+				&journal.entry.score_type,
+				JournalEntryOf::<T> {
+					entry: journal.entry.clone(),
+					digest: journal.digest,
+					block: block_number,
+				},
+			);
+			<JournalHashes<T>>::insert(&journal.digest, ());
+			<TidEntries<T>>::insert(
+				&journal.entry.tid,
+				&journal.entry.score_type,
+				&journal.entry.entity,
+			);
+
+			Self::deposit_event(Event::JournalEntry {
+				identifier,
+				entity: journal.entry.entity.clone(),
+				author: author.clone(),
+			});
+
+			Self::aggregate_score(&journal.entry);
```

Comment:
Should we be doing the aggregate score before event?  normally the practise is Event is raised at the end when all tasks are complete... Or you have designed it to be like this.